rotkehlchen/chain/ethereum/modules/convex/decoder.py

```diff
@@ -0,0 +1,131 @@
+from typing import Any, Callable, Dict, List, Optional, Tuple
+
+from rotkehlchen.accounting.structures.base import (
+    HistoryBaseEntry,
+    HistoryEventSubType,
+    HistoryEventType,
+)
+from rotkehlchen.assets.asset import EthereumToken
+from rotkehlchen.chain.ethereum.constants import ZERO_ADDRESS
+from rotkehlchen.chain.ethereum.decoding.interfaces import DecoderInterface
+from rotkehlchen.chain.ethereum.decoding.structures import ActionItem
+from rotkehlchen.chain.ethereum.modules.convex.constants import (
+    BOOSTER,
+    CONVEX_ABRAS_HEX,
+    CONVEX_POOLS,
+    CONVEX_VIRTUAL_REWARDS,
+    CPT_CONVEX,
+    CVX_LOCKER,
+    CVX_LOCKER_V2,
+    CVX_REWARDS,
+    CVXCRV_REWARDS,
+    REWARD_TOPICS,
+    TRANSFER_TOPIC,
+    WITHDRAWAL_TOPICS,
+)
+from rotkehlchen.chain.ethereum.structures import EthereumTxReceiptLog
+from rotkehlchen.chain.ethereum.utils import asset_normalized_value
+from rotkehlchen.types import ChecksumEthAddress, EthereumTransaction
+from rotkehlchen.utils.misc import hex_or_bytes_to_address, hex_or_bytes_to_int
+
+
+class ConvexDecoder(DecoderInterface):
+    @staticmethod
+    def _decode_convex_events(
+            tx_log: EthereumTxReceiptLog,
+            transaction: EthereumTransaction,
+            decoded_events: List[HistoryBaseEntry],
+            all_logs: List[EthereumTxReceiptLog],  # pylint: disable=unused-argument
+            action_items: List[ActionItem],  # pylint: disable=unused-argument
+    ) -> Tuple[Optional[HistoryBaseEntry], Optional[ActionItem]]:
+        amount_raw = hex_or_bytes_to_int(tx_log.data[0:32])
+        interacted_address = hex_or_bytes_to_address(tx_log.topics[1])
+
+        for event in decoded_events:
+            amount = asset_normalized_value(amount_raw, event.asset)
+            if (
+                event.location_label == transaction.from_address == interacted_address is False or
+                (event.counterparty != ZERO_ADDRESS and event.balance.amount != amount)
+            ):
+                continue
+            if (
+                event.event_type == HistoryEventType.SPEND and
+                event.event_subtype == HistoryEventSubType.NONE
+            ):
+                if event.counterparty == ZERO_ADDRESS:
+                    event.event_subtype = HistoryEventSubType.RETURN_WRAPPED
+                    event.counterparty = CPT_CONVEX
+                    if tx_log.address in CONVEX_POOLS:
+                        event.notes = f'Spent {event.balance.amount} {event.asset.symbol} as part of withdrawal from {CONVEX_POOLS[tx_log.address]} pool'  # noqa: E501
```

Comment:
Please don't use the "Spent" term for returning a pool token. I told you to look at other decoders and for example in compound it's:

`f'Return {redeem_tokens} {compound_token.symbol} to compound'`

So here you should do the same. So you return the pool token to the pool and as part of the same transaction withdraw the underlying tokens.

Reply:
Changed